Rewrite docs for `std::ptr` (Take #2)

ecstatic-morse opened the pull request:
Addresses #29371 and #36450.

This is a continuation of my work in [this PR](https://github.com/rust-lang/rust/pull/49767), which was merged prematurely.

The latest changes address all of @Gankro's concerns except the one regarding ordering of invariants for `copy_nonoverlapping`. The fact that memory regions must be nonoverlapping depends on the definition of the two memory regions above, and is also repeated in the function summary. 

@Gankro and @rkruppe pointed out that we don't have rigorously defined semantics for uninitialized memory, so all mention of it has been removed. The nomicon [implies that the functions in `std::ptr` work with uninitialized memory](https://doc.rust-lang.org/nomicon/unchecked-uninit.html), so requiring memory that is to be read be initialized is not an invariant. This conflicts with a strict interpretation of [the reference](https://doc.rust-lang.org/reference/behavior-considered-undefined.html), but resolving that is outside the scope of this PR.

Before this is merged, I need to resolve the following issues:

- [x] The new docs assert that `drop_in_place` is equivalent to calling `read` and discarding the value. Is this correct?
- [x] Do `write_bytes` and `swap_nonoverlapping` require properly aligned pointers?
- [x] Update docs for the unstable [`swap_nonoverlapping`](https://github.com/rust-lang/rust/issues/42818)
- [x] ~~Update docs for the unstable [unsafe pointer methods RFC](https://github.com/rust-lang/rfcs/pull/1966)~~ Will be done later.

In the meantime, I'm in desperate need of feedback. I'm also hosting a [rendered version of my changes](https://ecstatic-morse.github.io/rust/std/ptr/index.html) to make reviewing easier.

r? @steveklabnik

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/382967482) ([raw log](https://api.travis-ci.org/v3/job/382967482/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

[00:05:49] travis_fold:start:tidy
travis_time:start:tidy
tidy check
[00:05:49] tidy error: /checkout/src/libcore/ptr.rs:29: line longer than 100 chars
[00:05:51] some tidy checks failed
[00:05:51] 
[00:05:51] 
[00:05:51] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:05:51] 
[00:05:51] 
[00:05:51] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:05:51] Build completed unsuccessfully in 0:02:05
[00:05:51] Build completed unsuccessfully in 0:02:05
[00:05:51] make: *** [tidy] Error 1
[00:05:51] Makefile:79: recipe for target 'tidy' failed

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:05a4f4ee
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/383006117) ([raw log](https://api.travis-ci.org/v3/job/383006117/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:44:50] ....................................................i...............................................
[00:44:55] ........................................................................ii..........................
[00:45:01] ....................................................................................................
[00:45:07] ..................................................................................i.................
[00:45:09] iiiiiiiii...................................................
[00:45:09] 
[00:45:09] travis_fold:start:test_ui_nll
travis_time:start:test_ui_nll
Check compiletest suite=ui mode=ui compare_mode=nll (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
[00:45:56] ....................................................i...............................................
[00:46:00] .........................................................................ii.........................
[00:46:05] ....................................................................................................
[00:46:11] ..................................................................................i.................
[00:46:13] iiiiiiiii...................................................
[00:46:13] 
[00:46:13]  finished in 64.521
[00:46:13] travis_fold:end:test_ui_nll

---
[01:08:03] 
[01:08:03] running 1994 tests
[01:08:15] ....................................................................................................
[01:08:27] ....................................................................................................
[01:08:42] ......................................F.............................................................
[01:09:07] ....................................................................................................
[01:09:18] ....................................................................................................
[01:09:29] ....................................................................................................
[01:09:40] ....................................................................................................
---
[01:10:49] ....................................................................................................
[01:11:00] ....................................................................................................
[01:11:13] ....................................................................................................
[01:11:27] ....................................................................................................
[01:11:40] ..................................................F.................................................
travis_time:end:052e3bb8:start=1527134326161478552,finish=1527138659251255697,duration=4333089777145

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:01464bc1
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/383026522) ([raw log](https://api.travis-ci.org/v3/job/383026522/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:43:45] .....................................................i..............................................
[00:43:49] ........................................................................ii..........................
[00:43:55] ....................................................................................................
[00:44:01] ..................................................................................i.................
[00:44:03] iiiiiiiii...................................................
[00:44:03] 
[00:44:03] travis_fold:start:test_ui_nll
travis_time:start:test_ui_nll
Check compiletest suite=ui mode=ui compare_mode=nll (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
[00:45:04] ....................................................i...............................................
[00:45:10] ........................................................................ii..........................
[00:45:17] ....................................................................................................
[00:45:25] ..................................................................................i.................
[00:45:28] iiiiiiiii...................................................
[00:45:28] 
[00:45:28]  finished in 84.763
[00:45:28] travis_fold:end:test_ui_nll

---
[01:10:04] ....................................................................................................
[01:10:16] ....................................................................................................
[01:10:29] ....................................................................................................
[01:10:44] ....................................................................................................
[01:10:56] ................................................F...................................................
[01:11:30] ............................i.................................................................
[01:11:30] failures:
[01:11:30] 
[01:11:30] ---- ptr.rs - ptr::read (line 382) stdout ----
[01:11:30] ---- ptr.rs - ptr::read (line 382) stdout ----
[01:11:30] error: value assigned to `s2` is never read
[01:11:30]   --> ptr.rs:395:5
[01:11:30] 16 |     s2 = String::default();
[01:11:30]    |     ^^
[01:11:30]    |
[01:11:30] note: lint level defined here
[01:11:30] note: lint level defined here
[01:11:30]   --> ptr.rs:380:9
[01:11:30]    |
[01:11:30] 1  | #![deny(warnings)]
[01:11:30]    |         ^^^^^^^^
[01:11:30]    = note: #[deny(unused_assignments)] implied by #[deny(warnings)]
[01:11:30] thread 'ptr.rs - ptr::read (line 382)' panicked at 'couldn't compile the test', librustdoc/test.rs:325:17
[01:11:30] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:11:30] 
[01:11:30] 
---
[01:11:30] 
[01:11:30] error: test failed, to rerun pass '--doc'
[01:11:30] 
[01:11:30] 
[01:11:30] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "-j" "4" "--release" "--locked" "--color" "always" "--features" "panic-unwind jemalloc backtrace" "--manifest-path" "/checkout/src/libstd/Cargo.toml" "-p" "core" "--" "--quiet"
[01:11:30] 
[01:11:30] 
[01:11:30] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:11:30] Build completed unsuccessfully in 0:29:55
[01:11:30] Build completed unsuccessfully in 0:29:55
[01:11:30] make: *** [check] Error 1
[01:11:30] Makefile:58: recipe for target 'check' failed

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:00d31ac1
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Gankro commented:
Haven't had a chance to read this over, but just popping in to say:

* thanks for perserving on this!
* I believe it's reasonable to make the claims/requirements you state in your open issues

ecstatic-morse commented:
I'm pretty sure y'all have better things to do :). Right now the (no longer unstable) pointer methods are just copy-pasted versions of the old docs with a few notes added regarding argument ordering. I can copy them over, preserving the notes in a later PR, or merge them into this one. This thing is already pretty massive though, so I'm inclined to wait.

ecstatic-morse commented:
Oh and if the style is acceptable, I'll do `swap_nonoverlapping` this evening.

shepmaster commented:
Ping from triage, @steveklabnik !

shepmaster commented:
@ecstatic-morse your PR still has `[WIP]` in the title, but the reviewers seem generally happy. Can these two things be rectified?

bors commented:
:umbrella: The latest upstream changes (presumably #51310) made this pull request unmergeable. Please resolve the merge conflicts.

Gankro commented:
@shepmaster someone more intimate with the low-level details still needs to review this. I haven't had a chance and I don't know if I will.

steveklabnik commented:
Yes, same here.

ecstatic-morse commented:
I rebased to eliminate the merge commit I introduced by using github's merge conflict resolver.

Sorry, I'm sure there was a less messy way to do this.

rkruppe commented:
> A valid pointer is one that: [...]

This definition concerns me for a few reasons:

- It references LLVM rules (I know the reference does this as well, I don't like that either), which is pragmatic in that it gives one a good idea of what's most likely to broken by LLVM optimizations today, but it's no good for defining Rust rules
- It sounds exhaustive, but it really isn't -- and can't be, until we have unsafe code guidelines. This is to some degree unavoidable and by no means your fail, but strictly from a "how can we cover our asses" perspective I'd recommend some weasel words gesturing at the in-progress nature of some of these rules.

Besides these general points, I think the definition of "valid" needs to say more about Rust aliasing rules. The "general" LLVM aliasing rules are mentioned, but those are far weaker (the reference vaguely gestures at LLVM's "scoped noalias" to kind of cover how safe references behave). The upshot is: you can't really break the rules of safe references with raw pointer accesses. For example:
- it's UB to mutate memory while a shared borrow to it exists (unless it's in an `UnsafeCell`)
- while memory is mutably borrowed, it's UB to read or write the same memory through a different, unrelated pointer

Of course, the crux is defining what certain key phrases ("while a borrow exists", "unrelated pointer", etc.) mean exactly but even without unsafe code guidelines it's better to give vague guidance than omit it entirely.

ecstatic-morse commented:
@RalfJung, since you're working to formalize rust's memory model, do you have any thoughts on the proposed wording? I agree with @rkruppe that it's out of scope to exhaustively define what pointers are valid when passed to, for example, `ptr::write`, and I need to do a better job of pointing that out. However, I do believe that having a non-exhaustive list of conditions would be an improvement on the existing docs.

RalfJung commented on src/libcore/ptr.rs:
```diff
@@ -10,9 +10,29 @@
+//!
+//! * The pointer is not null.
+//! * The pointer is not dangling (it does not point to memory which has been
+//!   freed).
```
I'd strengthen this to something like "the pointer points to a live allocation". Arguably, `0x8 as *const u32` does not "point to memory which has been freed", but still it's not a valid pointer.

RalfJung commented on src/libcore/intrinsics.rs:
```diff
@@ -962,59 +962,129 @@ extern "rust-intrinsic" {
+    ///
+    /// * `src.offset(count-1)` must be [valid]. In other words, the region of
+    ///   memory which begins at `src` and has a length of `count *
+    ///   size_of::<T>()` bytes must belong to a single, live allocation.
```
The part after "in other words" says way more than the the first part! Also, Why is "`src` must be valid" and "`src.offset(count-1)`" split into two conditions? I see one condition here, which is "`src.offset(i)` must be valid for reading `size_of::<T>()` bytes for `0 <= i < count`", or, equivalently, "`src` must be valid for reading `size_of::<T>() * count` bytes`.

RalfJung replied:
I now see in your comments in the thread that you argue it would otherwise violate aliasing rules. That seems plausible but unnecessarily contorted for a definition.

Given that validity has to be defined for a size anyway, I'd just change this to require validity of the entire affected memory range at once (i.e. length `size_of::<T>() * count`).

ecstatic-morse replied:
I think I got a bit to clever here :) I was trying to rely exclusively on the definition of validity to define invariants over arrays as well as single accesses. The notion of a "single, live allocation" doesn't appear anywhere in the reference to my knowledge, which is why it went in "in other words".

As I note in my later comment, I think that making pointer validity a function of access size overcomplicates things a bit, and that the type of the pointer should fully specify the number of bytes being accessed when reading or writing.

From this point of view, your `src.offset(i)` for all `i` in `0..count` works formulation nicely, although it still references offset, which is maybe a bit opaque.

RalfJung replied:
> As I note in my later comment, I think that making pointer validity a function of access size overcomplicates things a bit, and that the type of the pointer should fully specify the number of bytes being accessed when reading or writing.

In my eyes, making it depend on the whole *type* makes things more complicated. With your proposal, validity is a function of the pointer and its type. A type contains much more information than just a size! With my proposal, it is crystal clear that the size of the only part that actually matters. Plus, for functions where the type alone does not suffice, we don't have to do awkward things like use `.offset` in our definition and rely on that function to be undefined when crossing allocation boundaries.

ecstatic-morse replied:
Does this mean I would need to qualify that, for example, `ptr::read` requires that its argument be valid for a read of at least size_of::<T>() bytes? This is what I meant by overcomplicates, not that it wasn't necessary for a formal definition, but it seems redundant when explicitly stated in the documentation.

Simply declaring that when the size of an access for `*const T` is not explicitly stated, we mean valid for `size_of::<T>()` bytes would work.

RalfJung replied:
I like this!

RalfJung commented on src/libcore/ptr.rs:
```diff
@@ -10,9 +10,29 @@
+//!
+//! Many functions in this module take raw pointers as arguments and dereference
+//! them. For this to be safe, these pointers must be valid. A valid pointer
+//! is one that satisfies **all** of the following conditions:
```
Validity of a pointer for an access is a property of the pointer, the size of the access and the kind of the access (reading or writing). That should be taken into account both by this definition and by every place that references it.

ecstatic-morse replied:
Interesting, the old PR used to differentiate between reads and writes, but only by requiring that reads must be done from initialized memory. I had to remove this, because it's not something we're ready to commit to, although it's currently suggested by the reference. I should have split validity into "valid for reads" and "valid for writes".

I view the size of the access as defined by the pointer type, so a `*const u32` can be valid for reads,(implying that the pointed-to region of memory is at least four bytes long), and/or valid for writes.

RalfJung commented on src/libcore/ptr.rs:
```diff
@@ -10,9 +10,29 @@
+//! * The pointer is not null.
+//! * The pointer is not dangling (it does not point to memory which has been
+//!   freed).
+//! * The pointer satisfies [LLVM's pointer aliasing rules].
```
This is not very helpful without further explanation on which LLVM aliasing annotations are emitted. That, in turn, may actually change.

I'd expect this to say, at least, that for a pointer to be valid for reading, there must be no live mutable references overlapping with the accessed memory region that this pointer is not based on (as defined in the LLVM rules).  For a pointer to be valid for writing, there must moreover be no live shared references overlapping with the accessed memory region, ignoring memory covered by an `UnsafeCell`.

That may sill not be sufficient, but it's a good start.

RalfJung replied:
I also agree with what @rkruppe wrote, this should contain an explicit note that these rules are still work-in-progress and hence subject to change and if there is any doubt, people should assume that what they are doing is probably UB.

RalfJung replied:
Which sections do you plan to link to?

ecstatic-morse replied:
Here's a list of possible candidates:
* [Dereferencing Raw Pointers in book](https://doc.rust-lang.org/book/second-edition/ch19-01-unsafe-rust.html#dereferencing-a-raw-pointer)
* [Undefined Behavior in reference](https://doc.rust-lang.org/reference/behavior-considered-undefined.html)
* [Aliasing in nomicon](https://doc.rust-lang.org/nomicon/aliasing.html)
* [Uninitialized Memory in nomicon](https://doc.rust-lang.org/nomicon/unchecked-uninit.html)
* [LLVM Pointer Aliasing](https://llvm.org/docs/LangRef.html#pointer-aliasing-rules): cited by the reference, and the only real resource for 

I propose linking to the first two. The first one contains some useful general info, and the second one references (`&mut`) aliasing, uninitialized memory and (C-style) pointer aliasing.

I'd like to defer further definitions to a later PR (this one is already suffering from scope creep), which would add a nomicon page discussing the validity of pointer accesses:

* In a vacuum, pointer validity depends on three factors, the value of the pointer, the size of the access, the type of the access (read or write).
* Additionally, a pointer is only valid for an access of a given size if it does not alias with a `&mut` reference (not entirely sure about the subtleties of this one).
* Reading or writing to an invalid pointer is undefined behavior
* A pointer is valid for reads if it points to uninitialized memory, but no assumptions can be made about the resulting value, or the results of any operations on the resulting value. Notably one cannot assume that it remains the same across multiple uses. For example (`x ^ x == 0`) may not be `true`, and `x & 0 == 0` may not be `true` if`x` was the result of a read of uninitialized memory. ( believe that this is stricter than LLVM's `undef`--for which the result of the first is not defined, but the result of the second is defined--but more relaxed than the current reference, which states that any read of uninitialized memory is UB. This may require a proper PR.
* The result of the casts `&x as *const _` and `&mut x as *mut _` is always a valid pointer (This is only true if creating a reference from an invalid pointer is already undefined behavior).
* Some stuff about pointer arithmetic and memory mapped I/O, which I am not currently ready to think about :)

RalfJung replied:
I'd add a reference to the aliasing page in the nomicon, but I believe you have such a reference included currently?

I see uninitialized memory as being orthogonal to what you are discussing here -- "uninitialized" is just yet another "value" that can be read (IOW, a byte is either a sequence of 8 bits, or "uninitialized"). This agrees with your characterization. It means that in principle, unintiailized memory does not have to be mentioned at all in the pointer validity discussion -- though it's probably worth clarifying this point as this question will come up.

RalfJung commented:
This is great! I'm always in favor of spelling out requirements more explicitly.

I haven't yet looked at every function documented here in detail, but there's one thing I would change in the overall organization: Most functions come with two requirements for the pointer, one saying "must be aligned" and one (or several, but see my comment) saying "must be valid". I think that's a good split, but I'd move the "non-null" requirement over to where it says "must be aligned".

The rationale is that being "valid for an access" is a property that depends on the size of the access. For example `0x8 as *const _` is valid for a read of length 0, but not for a read of length 1. So I'd put everything size-related into the definition of "valid for an access (of a given size)". Notably, that definition then *trivially holds* if the size is 0. All remaining conditions that have to hold even for 0-length accesses (non-null and aligned) should then be put separately into the function docs.

Now, it is a good question whether the pointer needs to be non-null and aligned if the access has size 0. Even the LLVM docs are not very clear on that for their intrinsics. But I feel it is better to say now that we require these properties to always hold and relax later, than go the other way around. That said, I seem to remember from my miri experiments that there is (or was) actually code in libstd that violates this for `copy` (namely `vec::IntoIter::next`) and `write_bytes` (namely `HashMap`) -- the intrinsics are called with size 0 and a pointer that is either not aligned or NULL.

ecstatic-morse commented:
Thanks for your comments @RalfJung, I have time over the weekend to make some changes.

ecstatic-morse commented:
@RalfJung 

The latest commits address some of your concerns. The `offset` based requirements on `copy` and friends have been removed in favor of something like "dst must be valid for a write of `count * size_of::<T>()` bytes". Also, each use of a pointer now states whether it must be valid for reads or writes (this is redundant for `*const`, but it's like three extra words). The module level docs have been changed to reflect this new wording: they state that a pointer must be valid for a given operation, a read or write of some number of bytes. It asserts that where the size of the operation is omitted, it is equal to the size of the pointed-to value in bytes.

I'm still trying to avoid tackling some of the finer points such as pointers to ZSTs, reads of uninitialized memory, and the subtleties of aliasing. As far as I know, these are still somewhat unresolved.

Btw, thanks for taking the time to review this so thoroughly.

pietroalbini commented:
Ping from triage @steveklabnik! What's the status on this?

steveklabnik commented:
as per https://github.com/rust-lang/rust/pull/51016#issuecomment-394403938, @gankro and I are basically done with this PR. It seems @RalfJung has done an actual review, but maybe not checked @ecstatic-morse 's latest work.

@bors: delegate=ralfjung :)

bors commented:
:v: @ralfjung can now approve this pull request

steveklabnik commented:
@bors: r? @RalfJung

pietroalbini commented:
They can't actually be assigned to the issue since they're not in the org.

RalfJung commented on src/libcore/ptr.rs:
```diff
@@ -10,9 +10,49 @@
+//!
+//! There are two types of operations on memory, reads and writes. It is
+//! possible for a `*mut` to be valid for one operation and not the other. Since
+//! a `*const` can only be read and not written, it has no such ambiguity. For
```
This makes the types mean much more than they should, I think. One can cast freely and in safe code between `*mut` and `*const`. For all intents and purposes, these are the same type; `*const` just has a "lint" associated with it that you cannot write through it.

RalfJung commented on src/libcore/ptr.rs:
```diff
@@ -10,9 +10,49 @@
+//! a `*const` can only be read and not written, it has no such ambiguity. For
+//! example, a `*mut` is not valid for writes if a a reference exists which
+//! [refers to the same memory][aliasing]. Therefore, each function in this
+//! module will document which operations must be valid on any `*mut` arguments.
```
I do not understand this last sentence. Do you mean something like, "each operation will document which `*mut` arguments it requires to be valid"?

RalfJung commented on src/libcore/ptr.rs:
```diff
@@ -10,9 +10,49 @@
+//!
+//! For more information on the safety implications of dereferencing raw
+//! pointers, see the both the [book] and the section in the reference devoted
+//! to [undefined behavior][ub].
```
I think that we can actually spell out the requirements *besides aliasing* here. As in, can't we start this section with the stuff we *know* -- that the range covered by `[pointer, pointer+size)` must be part of a single allocated memory block, and that the size is implicitly given by the type when not stated explicitly? Then we can say that there are additional requirements about aliasing, and that those rules are not settled yet, referencing the nomicon.

ecstatic-morse replied:
I added some axioms regarding validity, casting a reference to a pointer results in a valid pointer, pointers to ZSTs are valid, that sort of thing, and linked to `offset` as a proxy for the single allocated memory block requirement. I also called out aliasing in that final paragraph.

RalfJung commented:
Sorry for the long silence!

---

I have not carefully reviewed every single method doc, but focused on the notion of "validity". Beyond the comments above, I am missing a discussion of null pointers. I think you should do one of the following:

* For each method, add a requirement that the pointer be non-NULL. (This can be merged with the requirement of being aligned where applicable.) In the definition of validity, add a clarification saying that *all* pointers are valid for reading and writing 0 bytes.
* In the definition of validity, add a requirement that the pointer be non-NULL. Add a clarification saying that pointers are valid for reading and writing 0 bytes iff they are non-NULL.

In particular, I don't think zero-sized accesses are that complicated and I think they are relevant enough to be mentioned here. We are, I think, making them maximally undefined (as in, even zero-sized accesses have to be non-NULL and aligned), so we can always weaken that later if people complain.

I prefer the first variant because I like the fact that this makes validity hold trivially for zero-sized accesses. However, it leads to non-NULLness being repeated in plenty of places, so others might prefer the second variant. ;)

ecstatic-morse commented:
Sorry, I've been busy lately. It's a holiday in the U.S., so I'll pick this up again today.

ecstatic-morse commented:
I went with the second variant. I should have mentioned this several comments ago, but I don't believe that null pointers need to be a special case. Basically, we have two axioms, all pointers are valid for reads and writes of size zero, and a null pointer is *never* valid. These two axioms intersect when you have a null pointer to a ZST. Eventually, Rust will need to decide which axiom supersedes the other.

However, I don't think this PR is the place to settle this question, and propagating the non-null requirement everywhere is more committal than I'd like. Additionally, I think it's actually clearer to say "a null pointer is never valid" and "all pointers (except a null pointer) are valid for operations of size 0", than introduce "non-nullness" as an orthogonal concept to validity.

RalfJung commented:
> I think it's actually clearer to say "a null pointer is never valid" and "all pointers (except a null pointer) are valid for operations of size 0", than introduce "non-nullness" as an orthogonal concept to validity.

Fine for me. This is also the conservative choice, I think (allowing fewer things).

RalfJung commented on src/libcore/ptr.rs:
```diff
@@ -10,9 +10,65 @@
+//! memory][aliasing]. The set of operations for which a pointer argument must
+//! be valid is explicitly documented for each function. This is not strictly
+//! necessary for `*const` arguments, as they can only be used for reads and
+//! never for writes.
```
I am fine with the content of this, but the way this is phrased It may be just me but reading this paragraph leaves me pretty confused. For example, the "set of operations" can not be larger than 2, ever; and I also never find myself wondering "what is the set of operations that this pointer is valid for". I rather wonder "can I read through this"? Also, I do not think it is possible for a pointer to be valid for writing, but not valid for reading -- so "any combination" is not really possible.

So, I'd frame this entire thing by starting out saying there are *two* notions of validity: Valid for reading, and valid for writing. Every pointer operation specifies which kind of validity it requires. I'd put all of that into the first paragraph, and *not* talk about how Rust (typo: rust) doesn't yet have a fully defined model there. So, the first paragraph sets the stage and introduced the distinction between read and write validity, and does nothing else.

The second paragraph can then lead by saying that unfortunately, Rust doesn't yet have a fully defined memory model and hence the precise set of rules is not known yet. However, there are some things we do know -- e.g., if an `&mut` exists *and is not currently reborrowed* (that restriction is important), then any other aliasing pointer is not valid for reading nor writing. And so on.

It might even be better to move this paragraph down; the sentence right before your enumeration also touches the same topic so that would flow nicely.

RalfJung commented on src/libcore/ptr.rs:
```diff
@@ -10,9 +10,65 @@
+//! are a few rules regarding validity:
+//!
+//! * The result of casting a reference to a pointer is valid for as long as the
+//!   underlying object is live.
```
I'd say something like "a pointer obtained by casting a reference"; had to read this two times to parse it. ;)

Also, this is not true. The pointer obtained from the cast is valid immediately after the cast. However, using the reference again will invalidate the pointer, and other things may also invalidate it. For example:
```rust
fn foo(x: &mut u32) {
  let y = x as *mut;
  *x = 13;
  // Now y is NOT valid any more because `x` got "re-activated"
}
```

pietroalbini commented:
Ping from triage @ecstatic-morse! It's been a while since we heard from you, will you have time to work on this again?

pietroalbini commented:
Thank you for this PR @ecstatic-morse! Unfortunately we haven't heard from you on this in a while, so I'm closing the PR to keep things tidy. Don't worry though, if you'll have time again in the future please **reopen** this PR, we'll be happy to review it again!

RalfJung commented:
I might take over this PR, if I find time.

RalfJung commented:
Mostly for my own future reference -- a relevant issue in LLVM's bugtracker: https://bugs.llvm.org/show_bug.cgi?id=38583
